### ISSUE TITLE
pcs_resource: remove meta_attributes element ... (Issue #10)

### ISSUE THREAD

**Issue by Stevmann** [OndrejHome/ansible.pcs-modules-2#27]
Hello,

In my situation I have disabled resources on passive site of application. Resources are re-enabled after re-running the role, which creates those resources. The issue is probably here:

                    for elem in list(resource):
                        if elem.tag == 'meta_attributes' and len(list(elem)) == 0:
                            resource.remove(elem)
                    rc, diff = compare_resources(module, resource, clean_resource)

"meta_attributes" tag is not removed, so diff is always true, so resources are recreated, so enabled.
Thus condition "len(list(elem)) == 0" is not correct, or even unnecessary?
Thank you for advice.

S

**Comment by OndrejHome** [comment]
Hi @Stevmann, Thank you for question and for using pcs-modules-2.

Could you please provide the example playbook/task and state of the cluster ('pcs status') before and after you run that playbook?
also please provide the configuration of the resource before the playbook is run and mask any sensitive data from it ('pcs resource show resource-name').

The code you mention is intended only for cases where "meta_attributes" tag exists and doesn't have any values. In such case it can detect difference in CIB while there is no practical difference (this was caused by code in 'pcs' that should be now fixed, but previously could leave stuff in this state). 
I wonder how you define the resource that gets re-enabled after re-running. My guess will be that you define resource without specifying in the 'options' the "meta Target-Role=Stopped" which is a way how to create a disabled resource. pcs_resource module tries to create a resource that you have specified in temporary file and then it compares its definition with what is running in cluster. If the resource in pcs_resource is defined without special options that would mean that desired state or resource is running and running taks with this would update the running version of resource (if it exists) to match this state. Point here is that "resource is disabled" is and an Meta Attribute of  resource ('target-role').

**Comment by Stevmann** [comment]
Hello,

Thank you for your quick reply and sorry for my late one. It is not easy for me to send requested informations, but I will try to describe it another way. In case you will still need them, I will try to figure out something.

First,to clarify re-enablig and re-running... :)
On passive site of datacenter, there is application with pacemaker. But resources are disabled. So pcs status shows:

 <some_resource>  (ocf::*:*): Stopped (disabled)

When we need to make this site active (in case of issue on active site), we have to manually do "pcs resource <resource> enable" (among other things)... But from time to time, to keep environments aligned (versions of binaries,clean up mess,etc.) we run ansible roles on them. So also the playbook, which consists also of the task,that creates resources. (We use the same also,when we build new environments.) Something like:

- name: Create Pacemaker <some_resource> Resource
  pcs_resource:
    name: <some_resource>
    resource_class: ocf
    resource_type: ocf:*:*
    options: servicename=<some_resource> op monitor interval=30s timeout=60s
  run_once: true

In our case, this task (module) will enable our disabled resource because:
According to 'pcs_resource.py' it will create clean resource to compare it with the existing one.
So it does actually 'diff' of two temporary files (resource configurations).
And if we look closer at those files,in the existing resource, meta_attributes tag is still present,because like you wrote, it will remove only empty meta_attributes.
So it looks like this:

    <primitive class="ocf" id="<some_resource>" provider="*" type="*">
        <meta_attributes id="<some_resource>-meta_attributes">
          <nvpair id="<some_resource>-meta_attributes-target-role" name="target-role" value="Stopped" />
        </meta_attributes>
        <instance_attributes id="<some_resource>-instance_attributes">
          <nvpair id="<some_resource>-instance_attributes-servicename" name="servicename" value="<some_resource>" />
        </instance_attributes>
        <operations>
          <op id="<some_resource>-monitor-interval-30s" interval="30s" name="monitor" timeout="60s" />
          <op id="<some_resource>-start-interval-0s" interval="0s" name="start" timeout="300" />
          <op id="<some_resource>-stop-interval-0s" interval="0s" name="stop" timeout="20" />
        </operations>
    </primitive>

Which is different (contains meta_attributes - resource stopped) from clean_resource, so it recreates it and so enables it. Unfortunately "--disable" option for creating disabled resources is not usable, because from the same reason it can disable started resource.
I agree,maybe our situation I described above is somewhat specific or unusual, but anyway,standard behaviour of pacemaker if you run "pcs resource create ..." for existing resource (started or disabled) is:

Error: '<some_resource>' already exists

Anyway, it's great to have this module. ;)
Regards,
S

**Comment by OndrejHome** [comment]
Hi, thank you for more detailed explanation. I understand it that youe use case is to check resource configuration ignoring the "enabled/disabled" state. The 'pcs_resource' module is not capable of ignoring this. I have on my mind 2 ways how to address this:

1. add variable to 'options' variable of 'pcs_resource' that will contain either nothing or "meta target-role=stopped" depending on which cluster you run this. In cluster that should have active resources it will be empty, in clustet where you wanna disabled resources it will contain this option that makes resource disabled. This is something that would not require change to 'pcs_resource' module.
2. I can have a look on how to implement to 'pcs_resource' module option to ignore either this use case (enabled/disabled) resource or more generic state. I can imagine having something like "state: defined" that would ignore resource state for example.

I can have a more technical look at this either over this weekend or next week. You can try the approach 1 to see if that would address your use case or wait for me. In overall this is interesting use case as it look to me as check tool to see if anything (but state of resource) has changed. In any case I would either document the approach 1 if it works well in 'pcs_resource' EXAMPLES or implement a change to 'pcs_resource' module. Thank you and please let me know if 1 worked for you or if you would rather prefer the 2. In case of 2 I would justvlike to ask you for testing out the code once implemented to confirm that it works for you before it is merged into 'master'.

**Comment by Hoijima** [comment]
Hi, this topic is also vital to me, but for the most part i had gone from just creating and deleting resources just to a plain management of them.
Currently i don't try to interface with the meta attributes, because it has proven to be not reliable in terms of the end result.

I modified the the pcs_resource this way.

i added 2 new states of the resource : enable, disable
that directly tells pcs to enable or disable the resource, just via the command

pcs resource enable \ pcs resource disable

using refreshing and cleaning up before it's start.
In the end it looks something like this,
The module will try to enable or disable the resource, wait a bit (parametrized) and if it's starting or stopping will report a changed status, if anything else will just fail.
This approach proven to be more efficient in terms of the expected results.
But may be a bit out of place(in terms of ansible philosophy probably) due to the fact i made the functionality directly in pcs_resource. I think i would move the functinality some where like pcs_manage, but in terms of cmd usage, enable\disable IMHO is better than meta.

**Comment by OndrejHome** [comment]
Hi @Hoijima ,

Thank you for additional information. Based on that lets move the discussion to new Issue #28 .
I will post there my thoughts shorty.

**Comment by OndrejHome** [comment]
@Stevmann ,

Thank you for patience. I have one code in development that give ability to specify list of ignored meta attributes when comparing with existing resource. Example below.

Once I put it into better shape for commit I will let you know so you can test it. But in short: It looks that this will be possible without big change in code, which is good. :)

~~~
diff --git a/library/pcs_resource.py b/library/pcs_resource.py
index b27e586..8cba7d9 100644
--- a/library/pcs_resource.py
+++ b/library/pcs_resource.py
@@ -233,6 +233,7 @@ def run_module():
             force_resource_update=dict(default=False, type='bool', required=False),
             cib_file=dict(required=False),
             child_name=dict(required=False),
+            ignored_meta_attributes=dict(required=False, type='list', elements='str', default=[]),
         ),
         supports_check_mode=True
     )
@@ -245,6 +246,7 @@ def run_module():
         module.params['child_name'] = resource_name + '-child'
     child_name = module.params['child_name']
     resource_options = module.params['options']
+    ignored_meta_attributes = module.params['ignored_meta_attributes']
 
     if state == 'present' and (not module.params['resource_type']):
         module.fail_json(msg='When creating cluster resource you must specify the resource_type')
@@ -386,10 +388,31 @@ def run_module():
             clean_resource = find_resource(cib_clean_resources, resource_name)
 
             if clean_resource is not None:
+                #TODO refactor into something smaller or fucntion
+                result['ignored'] = ignored_meta_attributes
+                # remove the ignored_meta_attributes from resource
+                for elem in list(resource):
+                    if elem.tag == 'meta_attributes' and len(list(elem)) > 0:
+                        for nvpair in list(elem):
+                            if nvpair.tag == 'nvpair' and nvpair.attrib.get('name') in ignored_meta_attributes:
+                                elem.remove(nvpair)
                 # remove the meta_attribute element from original cluster cib when empty to make comparison clean - Issue #10
                 for elem in list(resource):
                     if elem.tag == 'meta_attributes' and len(list(elem)) == 0:
                         resource.remove(elem)
+
+                # remove the ignored_meta_attributes from clean_resource
+                for elem in list(clean_resource):
+                    if elem.tag == 'meta_attributes' and len(list(elem)) > 0:
+                        for nvpair in list(elem):
+                            if nvpair.tag == 'nvpair' and nvpair.attrib.get('name') in ignored_meta_attributes:
+                                elem.remove(nvpair)
+                # remove the meta_attribute element from cluster cib when empty to make comparison clean
+                for elem in list(clean_resource):
+                    if elem.tag == 'meta_attributes' and len(list(elem)) == 0:
+                        clean_resource.remove(elem)
+
+                # compare the existing resource in cluster and simulated clean_resource
                 rc, diff = compare_resources(module, resource, clean_resource)
                 if rc == 0:
                     # if no differnces were find there is no need to update the resource
~~~

~~~
---
- hosts: all
  remote_user: root
  roles:
    - OndrejHome.pcs-modules-2
  tasks:
    - pcs_resource:
        name: 'test'
        resource_type: 'ocf:pacemaker:Dummy'
        options: 'meta target-role=Stopped'
        ignored_meta_attributes:
          - 'target-role'
          - 'whatever-other-meta-attribute'
      delegate_to: 192.168.22.31
      run_once: true
~~~

**Comment by OndrejHome** [comment]
Hi @Stevmann ,
In `develop` branch I have pushed changes for this Issue that allows specifying which meta attributes can be ignored if desired in `pcs_resource`. 
Could please let me know if this works for your use case here and also if you have any additional questions or feedback for this feature?
(You can check updated `ansible-doc pcs_resource` on how to use this for ignoring state of resource).

I will wait for reply from you on testing this. In case there is no reply I will consider that this is OK and I will merge this to `master` branch when doing next release. If you would like to test this but need more time please let me know. I always appreciate when the feature is tested by original requester :)

**Comment by Stevmann** [comment]
Hi @OndrejHome ,

Thank you very much for your effort. It looks really good. I am sorry for delays in my communication. :(
I believe, that I will have time to do some tests during weekend.

Regards

**Comment by OndrejHome** [comment]
Thank you @Stevmann , I will wait here for feedback from you. No need to hurry I expect to be doing more thing during next week/weekend. If you manage to test this out or find any issue with it feel free to leave comment ;)

**Comment by Stevmann** [comment]
Hi @OndrejHome 
I confirm, that your fix works for me. 
Thank you very much and wish you all the best.
S

**Comment by OndrejHome** [comment]
Hi @Stevmann , 
Thank you for confirmation and reporting this issue in first place! Changes are now merged in `master` branch and new tag `21.0.0.` was published. Also, all the best to you :)

Ondrej
